### DISCUSSION TITLE
VideoEffect doesn't run execute function

### DISCUSSION THREAD

**Discussion by abhinandan-chakraborty** [HaishinKit/HaishinKit.swift#1504] [Q&A] [answered]
Hi, first of all, thank you very much for this amazing library and keeping it up-to-date and supporting so many features out of the box. I have been playing around with this library for a couple of days but I can't seem to figure out why VideoEffect "execute" function never gets called. Here is an example:

```swift
private class ContentModel: ObservableObject, IOStreamObserver, IOStreamDelegate {
    @Published var stream:RTMPStream?
    private let connection = RTMPConnection()
    
    init() {
            let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker, .allowBluetooth])
            try session.setActive(true)
                
            stream = RTMPStream(connection: connection)
            stream?.videoOrientation = .portrait

            stream?.sessionPreset = .hd1280x720
            stream?.videoSettings.profileLevel = kVTProfileLevel_H264_Baseline_AutoLevel as String
            stream?.videoSettings.videoSize = CGSize(width: 720, height: 1280)
            stream?.videoSettings.bitRate = 4 * 1000 * 1000
            stream?.frameRate = 30

            stream!.attachAudio(AVCaptureDevice.default(for: .audio)) { _, error in
              if let error {
                  print("errr 1: \(error.localizedDescription)")
              }
            }
                        
            stream!.attachCamera(AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back)) { videoUnit, error in
              if let error {
                  print("errr 2: \(error.localizedDescription)")
              } 
            }
            
            stream?.addObserver(self)
            stream?.delegate = self
            
            let _ = stream?.registerVideoEffect(SepiaEffect()) //here attempts to use Video effect
            
        } catch(let err) {
            print("error :\(err)")
        }
    }
    
    
    //on button click
    public func startStream() {
        connection.connect("rtmp://192.168.1.38/live")
        stream?.publish("hello")
    }
}
```
and now for the my video effect class:
```swift
final class SepiaEffect: VideoEffect {
    let filter = CIFilter.sepiaTone()
    
    override init() {
        print("sepiaeffect ") // this gets called successfully
    }

    override func execute(_ image: CIImage, info: CMSampleBuffer?) -> CIImage {
        print("called here") //never gets called
        filter.inputImage = image
        filter.intensity = 0.9
        return filter.outputImage!
    }
}
```

now talking about my true intention, I don't really need to use CIFilter in my actual project, as I just want  to access `CMSampleBuffer` and send it to `VTDecompressionSessionDecodeFrame` to decode multi-frame MV-HEVC and then process it and send it back as either a modified `CMSampleBuffer` or `CIImage` (If I use VideoEffect).

There is another way where I can access `CMSampleBuffer`:
```swift
func stream(_ stream: HaishinKit.IOStream, track: UInt8, didInput video: CMSampleBuffer) {
  //do some processing and then:
  stream?.append(newSampleBuffer, track: track)
|
```
but this method causes the `MTHKView` to flicker ( assuming append makes the stream add same buffer instead of replacing original buffer with the new `CMSampleBuffer` object ). I also tried to use another `RTMPStream` object that would publish the stream but there the stream is very unstable. All in all, I think `VideoEffect` sounds like the right way to deal with this situation, but I am not sure why `execute` is never called.

**Comment by shogo4405** [comment, marked as answer]
If you are using version 1.9.x, you need to specify the following value. The default is [.passthrough](https://docs.haishinkit.com/swift/1.9.1/Structs/IOVideoMixerSettings/Mode.html#/s:10HaishinKit20IOVideoMixerSettingsV4ModeO11passthroughyA2EmF).

```stream.videoMixerSettings.mode = .offscreen```

**Reply by abhinandan-chakraborty** [reply]
Ah, what a beauty! It works so good. I just had to add these lines:
```swift
   stream?.videoMixerSettings.mode = .offscreen
   stream?.screen.startRunning()
   stream?.registerVideoEffect(SepiaEffect())
```

Just one last question, Is it possible to publish these filtered streams to the server but without displaying these effects on top of `MTHKView`?

Thank you for so fast response.

**Reply by shogo4405** [reply]
I did not design it that way. However, it is possible to display a raw CMSampleBuffer in MTKView. In this case, calling hkView.attachStream is not necessary.

```swift
func stream(_ stream: HaishinKit.IOStream, track: UInt8, didInput video: CMSampleBuffer) {
  //do some processing and then:
  if track == 0 {
      hkView?.enqueue(sampleBuffer)
  }
|
```

**Reply by abhinandan-chakraborty** [reply]
Wow, this is just so cool. It works perfectly fine. Thank you so much for your help!
